order by implementation

onurkuruu opened the pull request:
*Issue #47 *

* Support for `NULLS FIRST`, `NULLS LAST` keywords
* Support for ordering `null/missing`, `bool`, `number`, `date/timestamp/time`, `string` types
* A bug fixed which `SqlParser` cannot parse if the statement contains `ORDER BY` which comes after `GROUP BY`
* Support for ordering the results;
	* Normal results
	* Aggregated results without grouping (GROUP BY)
	* Grouped results


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

alancai98 commented on lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt:
```diff
@@ -1982,6 +2002,35 @@ internal class EvaluatingCompiler(
             CompiledGroupByItem(alias.text.exprValue(), uniqueName, compileAstExpr(it.expr))
         }
 
+    private fun compileOrderByExpression(sortSpecs: List<PartiqlAst.SortSpec>): List<CompiledOrderByItem> =
+        sortSpecs.map {
+            it.orderingSpec
+                ?: errNoContext(
+                    "SortSpec.orderingSpec was not specified",
+                    errorCode = ErrorCode.INTERNAL_ERROR,
+                    internal = true
+                )
+
+            it.nulls
+                ?: errNoContext(
+                    "SortSpec.nulls was not specified",
+                    errorCode = ErrorCode.INTERNAL_ERROR,
+                    internal = true
+                )
+
+            val orderSpec = when (it.orderingSpec) {
+                is PartiqlAst.OrderingSpec.Asc -> OrderingType.ASC
+                is PartiqlAst.OrderingSpec.Desc -> OrderingType.DESC
+            }
+
+            val nullsSpec = when (it.nulls) {
+                is PartiqlAst.NullsSpec.NullsFirst -> NullsType.FIRST
+                is PartiqlAst.NullsSpec.NullsLast -> NullsType.LAST
+            }
+
+            CompiledOrderByItem(orderSpec, nullsSpec, compileAstExpr(it.expr))
+        }
```
I think [NaturalExprValueComparators](https://github.com/partiql/partiql-lang-kotlin/blob/67e3adf452d1d411aaba71de2635e12bb25bca2e/lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt) can be used here with some slight modification to case on the `OrderingSpec`. Using that could remove the need for `CompareExtentions.kt` and provide the implementation for some of the type comparisons.

Also, `NaturalExprValueComparators` has a lot of tests in [NaturalExprValueComparatorsTest.kt](https://github.com/partiql/partiql-lang-kotlin/blob/67e3adf452d1d411aaba71de2635e12bb25bca2e/lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt), which can reduce the amount of tests needed to be written.

alancai98 commented on lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt:
```diff
@@ -0,0 +1,95 @@
+                "SELECT * FROM products ORDER BY price DESC, productId ASC LIMIT 1",
+                "<<{ 'productId': 3, 'supplierId': 10, 'categoryId': 21, 'price': 15.0, 'numInStock': 100 }>>"
+            ),
+            // should order by supplierId_nulls nulls last
+            EvaluatorTestCase(
+                "SELECT productId FROM products_sparse ORDER BY supplierId_nulls NULLS LAST, productId",
+                "<<{'productId': 1}, {'productId': 2}, {'productId': 3}, {'productId': 6}, {'productId': 7}, {'productId': 4}, {'productId': 5}, {'productId': 8}, {'productId': 9}, {'productId': 10}>>"
+            ),
+            // should order by supplierId_nulls nulls first
+            EvaluatorTestCase(
+                "SELECT productId FROM products_sparse ORDER BY supplierId_nulls NULLS FIRST, productId",
+                "<<{'productId': 4}, {'productId': 5}, {'productId': 8}, {'productId': 9}, {'productId': 10}, {'productId': 1}, {'productId': 2}, {'productId': 3}, {'productId': 6}, {'productId': 7}>>"
+            )
```
nit: may be helpful to include `supplierId_nulls` in the select projection to verify the results

alancai98 reviewed (changes requested):
Thanks for the contribution and putting together this PR! This is definitely a great start to the implementation! I noticed a few things that need addressed:
- changing default behavior when ordering spec (i.e. `ASC`/`DESC`) is not specified
- using `NaturalExprValueComparators` to perform the `ORDER BY` comparison between the different types
- some more tests (parsing, evaluation)
Other than that, only had some nits/typos.

onurkuruu commented:
Hi @alancai98, I have pushed a commit to the branch. I tried to use `NaturalExprValueComparators` to order the rows but I am not sure about the code changes if this is what you wanted. If this is the correct approach you wanted, I can add more test cases to `NaturalExprValueComparatorsTest` to cover new changes.

Here is the what the commit contains;

* fixed some typo
* added `asc/desc` support to `NaturalExprValueComparators`
* using `NaturalExprValueComparators` when ordering rows
* `getQueryThunk` returning list if order by specified otherwise bag
* added more test cases to `SqlParserTest` and `EvaluatingCompilerOrderByTests`

onurkuruu commented on lang/src/org/partiql/lang/ast/ast.kt:
```diff
@@ -1055,6 +1056,13 @@ enum class OrderingSpec {
 
+/** Nulls specification */
+enum class NullsSpec {
+    /** Represents */
```
I have added this with new commit.

onurkuruu commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -132,6 +132,7 @@ class SqlParser(
         ORDER_BY,
         SORT_SPEC,
         ORDERING_SPEC,
+        NULLS,
```
I have added this with new commit.

onurkuruu commented on lang/test/org/partiql/lang/errors/ParserErrorsTest.kt:
```diff
@@ -1134,6 +1134,90 @@ class ParserErrorsTest : SqlParserTestBase() {
         )
     }
 
+    @Test
```
Thanks!

onurkuruu commented on lang/test/org/partiql/lang/syntax/SqlParserTest.kt:
```diff
@@ -1696,7 +1696,8 @@ class SqlParserTest : SqlParserTestBase() {
                     (sort_spec 
                         (id rk1 (case_insensitive) (unqualified)) 
-                        (asc)))))
+                        (asc)
```
I have some more test cases which also covers `NULLS FIRST`/`NULLS LAST`

onurkuruu commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -577,8 +578,13 @@ class SqlParser(
+                                        ParseType.NULLS -> sortSpec(it.children[0].toAstExpr(), asc(), it.children[1].toNullsSpec())
+                                        else -> errMalformedParseTree("Invalid ordering expressions syntax")
+                                    }
+                                    3 -> sortSpec(it.children[0].toAstExpr(), it.children[1].toOrderingSpec(), it.children[2].toNullsSpec())
```
I have updated here as you defined above

onurkuruu commented on lang/src/org/partiql/lang/errors/ErrorCode.kt:
```diff
@@ -557,6 +557,22 @@ enum class ErrorCode(
+        override fun getErrorMessage(errorContext: PropertyValueMap?): String = ""
+    },
+
+    EVALUATOR_ORDER_BY_ENVIRONMENT_CANNOT_RESOLVED(
```
I have changed here with `EVALUATOR_ENVIRONMENT_CANNOT_BE_RESOLVED` since I have changed the logic of the code which use this error message.

alancai98 commented on lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:
```diff
@@ -55,9 +55,17 @@ import org.partiql.lang.util.isZero
+    NULLS_LAST_ASC(Order.ASC, NullOrder.LAST),
+    NULLS_LAST_DESC(Order.DESC, NullOrder.LAST);
+
+    /** Whether or not null values come first or last. */
```
typo: doc comment should talk about the ordering spec not null order spec

onurkuruu replied:
fixed with new commit

alancai98 commented on lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:
```diff
@@ -55,9 +55,17 @@ import org.partiql.lang.util.isZero
-enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Comparator<ExprValue> {
-    NULLS_FIRST(NullOrder.FIRST),
-    NULLS_LAST(NullOrder.LAST);
+enum class NaturalExprValueComparators(private val order: Order, private val nullOrder: NullOrder) : Comparator<ExprValue> {
```
nit (documentation): since we're adding another argument to the public constructor, could add a kdoc comment for the new parameter, `order`

onurkuruu replied:
fixed with new commit

alancai98 commented on lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt:
```diff
@@ -1998,6 +2025,92 @@ internal class EvaluatingCompiler(
+
+            it.nullsSpec
+                ?: errNoContext(
+                    "SortSpec.nulls was not specified",
```
typo: should say `SortSpec.nullsSpec`

```suggestion
                    "SortSpec.nullsSpec was not specified",
```

alancai98 commented on lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt:
```diff
@@ -0,0 +1,105 @@
+                "[{'sellerId': 2}, {'sellerId': 1}]"
+            )
+        )
+    }
```
This is a good start for the evaluation tests. I'll discuss further with the team about what level of coverage seems reasonable as `NaturalExprValueComparatorsTest` covers a lot of edge cases. Can consider adding more tests with
- `MISSING`, `NULL` and `MISSING`
- different data types (not just numbers)

onurkuruu replied:
I have added more test cases which covers different data types and edge cases

alancai98 reviewed (commented):
Changes since the initial commit look good. Identified a few nits/typos and there could be some more evaluation tests with other data types.

I think the approach using `NaturalExprValueComparators` looks reasonable. What other tests were you considering adding?

onurkuruu commented:
> Changes since the initial commit look good. Identified a few nits/typos and there could be some more evaluation tests with other data types.
> 
> I think the approach using `NaturalExprValueComparators` looks reasonable. What other tests were you considering adding?

I thought , I could add some tests to cover `ASC/DESC` feature of `NaturalExprValueComparators` to see if everything working as expected. Actually, EvaluatingCompilerOrderByTests covers `ASC/DESC` feature of `NaturalExprValueComparators` but I thought, It would be good also adding some test cases for `NULLS_FIRST_DESC` and `NULLS_LAST_DESC`.

alancai98 commented:
>I thought , I could add some tests to cover `ASC/DESC` feature of `NaturalExprValueComparators` to see if everything working as expected. Actually, EvaluatingCompilerOrderByTests covers `ASC/DESC` feature of `NaturalExprValueComparators` but I thought, It would be good also adding some test cases for `NULLS_FIRST_DESC` and `NULLS_LAST_DESC`. 

Yeah I think those would be good to add.

---

Regarding the tests for [EvaluatingCompilerOrderByTests.kt](https://github.com/partiql/partiql-lang-kotlin/blob/2b2a1f26951bee018af377c91d86832a65a635ba/lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt), we think more tests should be added for the end-to-end evaluator behavior (even though there are unit tests for [NaturalExprValueComparators](https://github.com/partiql/partiql-lang-kotlin/blob/2b2a1f26951bee018af377c91d86832a65a635ba/lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt)). Tests should ensure the ORDER BY ordering follows what's specified in [section 12.2](https://partiql.org/assets/PartiQL-Specification.pdf#page=44) of the spec. As a summary, should provide the following tests:
- tests between different data types
  - assuming `ASC` and `NULLS FIRST`, check end-to-end evaluation follows the order: unknowns (null, missing), bool, numbers, date, time, timestamp, text, LOB types, lists, structs, bags)
- edge case tests for values of the same data type
  - `false` before `true`
  - `nan` before `-inf`, then numeric values, then `+inf`
  - `timestamp`s compared by absolute point of time
  - `text` types compared by lexicographical ordering of Unicode scalar
  - `LOB` types follow their lexicographical ordering by octet
  - `list`s compared lexicographically based on comparison of elements, recursively
    - also, shorter arrays come first
  - `struct`s compared lexicographically based on sorted attributes (first by key then by value)
  - `bag`s compared as sorted lists

Feel free to use the data provided in the tests from [NaturalExprValueComparatorsTest.k](https://github.com/partiql/partiql-lang-kotlin/blob/2b2a1f26951bee018af377c91d86832a65a635ba/lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt#L28-L291). If it's any easier, you can also include the data directly in the from source rather than the evaluation session. E.g.:

```
SELECT * FROM [{ 'a': false }, { 'a': true }] ORDER BY a
```

onurkuruu commented on lang/resources/org/partiql/type-domains/partiql.ion:
```diff
@@ -217,14 +217,20 @@
         (product order_by sort_specs::(* sort_spec 1))
 
         // <expr> [ASC | DESC] ?
```
fixed with new commit

onurkuruu commented on lang/test/org/partiql/lang/syntax/SqlParserTest.kt:
```diff
@@ -1777,12 +1782,155 @@ class SqlParserTest : SqlParserTestBase() {
         """
     )
+
+    @Test
```
Thanks!

onurkuruu commented:
Thanks for your review @alancai98 , I have pushed a new commit for your comments.
* Added necessary kdoc comments 
* Fixed some typo
* Added more test cases which covers different data types and edge cases as defined at specifications for `EvaluatingCompilerOrderByTests` 
* Refactored `NaturalExprValueComparatorsTest` to cover desc feature of `NaturalExprValueComparators`

alancai98 commented on lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:
```diff
@@ -157,6 +170,13 @@ enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Compa
         }
     }
 
+    private fun <T> orderLeftRight(order: Order, left: T, right: T): Pair<T, T> {
```
Swapping the left and right side ordering is what we want when comparing between two of the **same** data types for `DESC`. But since we also want values of different types to be sorted in descending order, this approach won't work. To resolve this without repeating too much of the logic, we could break the old version of `compare` into its own function and return the inverse result.

See this gist with one approach to reverse the result of comparison for `DESC`: https://gist.github.com/alancai98/c69fc04f8c4c41177c67c8ffa064012c

onurkuruu replied:
I have committed the code from your gist and changed the tests a bit. Now we are comparing different data types with the following order (Nulls/Missing can be changed by ordering and nulls spec)
* bools, numbers, timestamp, text ... for ascending order
* bag, tuple, array, lob for descending order

alancai98 reviewed (commented):
@onurkuruu, thanks for adding so many tests! While going through them, I noticed that the semantics of `ORDER BY`'s `DESC` are not entirely correct when comparing between different data types. I tried to explain some of the reasoning and provide a fix in the comments.

Feel free to let me know if anything's confusing or if you have further questions!

alancai98 commented on lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt:
```diff
@@ -0,0 +1,278 @@
+                """SELECT * FROM [{'a': `{{"Z"}}`}, {'a': `{{"a"}}`}, {'a': `{{"A"}}`}, {'a': `{{"z"}}`}] ORDER BY a DESC""",
+                """[{'a': `{{"z"}}`}, {'a': `{{"a"}}`}, {'a': `{{"Z"}}`}, {'a': `{{"A"}}`}]"""
+            ),
+            // sorter array comes first (ASC)
```
typo: shorter
```suggestion
            // shorter array comes first (ASC)
```

alancai98 commented on lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:
```diff
@@ -104,8 +115,10 @@ enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Compa
         while (lIter.hasNext() && rIter.hasNext()) {
-            val lChild = lIter.next()
-            val rChild = rIter.next()
+            val (lChild, rChild) = when(order) {
```
Seems like GitHub actions builds aren't run for forks as mentioned in #561. When I build locally, this line caused a ktlint error. Fixable by running `./gradlew ktlintFormat` or applying the following suggestion:

```suggestion
            val (lChild, rChild) = when (order) {
```

alancai98 replied:
Similarly with L142.

alancai98 commented on lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:
```diff
@@ -126,11 +139,15 @@ enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Compa
         val pairCmp = object : Comparator<Pair<T, Int>> {
             override fun compare(o1: Pair<T, Int>, o2: Pair<T, Int>): Int {
-                val cmp = entityCmp.compare(o1.first, o2.first)
+                val (leftPair, rightPair) = when(order) {
```
From the ktlint error:

```suggestion
                val (leftPair, rightPair) = when (order) {
```

alancai98 commented on lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:
```diff
@@ -294,4 +311,22 @@ enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Compa
         throw IllegalStateException("Could not compare: $left and $right")
     }
+
+    // can think of `DESC` as the converse/result of `ASC`
```
Was a typo from my .gist:

```suggestion
    // can think of `DESC` as the converse/reverse of `ASC`
```

alancai98 reviewed (commented):
Only have a few small nits/style fixes. Should be ready to approve once addressed

onurkuruu commented:
@alancai98 thanks for your review again. It was pleasure to contribute with your support. What are the steps of releasing new features of  PartiQL? When we can use the PartiQL with order by support with official release?

alancai98 commented:
Thanks again for the contribution and the quick responses to feedback!

> What are the steps of releasing new features of PartiQL? When we can use the PartiQL with order by support with official release?

We typically publish new PartiQL versions through [maven](https://search.maven.org/search?q=a:partiql-lang-kotlin). There's currently no release schedule in place, but we can perform the release whenever a customer/user needs the latest version. Do you need this latest version released to maven?

onurkuruu commented:
Yes, we need. It would be great to have it.

alancai98 commented:
>Yes, we need. It would be great to have it.

Got it. We should be able to start the release process today. I'll followup here w/ any updates.

alancai98 commented:
@onurkuruu the latest release, [v0.6.0](https://github.com/partiql/partiql-lang-kotlin/releases/tag/v0.6.0-alpha), was published to [Maven](https://search.maven.org/artifact/org.partiql/partiql-lang-kotlin/0.6.0/jar) earlier today. Feel free to let us know if you have any issues using it!

onurkuruu commented:
Thanks @alancai98, I can use `Order By` now. I also want to add alias support for `Order By` and `Group By`. Currently, following queries are not working because of missing bindings.
* `SELECT a as b FROM table ORDER BY b`
* `SELECT a as b FROM table GROUP BY b`

Do you mind if I open a new PR to support for aliases, if can implement? or Do I need to open an issue for this?

alancai98 commented:
Hey @onurkuruu. Sorry have been away due to a bout with COVID. I think it would be good to have select aliases available to `GROUP BY` and `ORDER BY` clauses. I believe PostgreSQL, MySQL, and SQLite support this. I created a GitHub issue #571 from your previous comment.
